sycl/source/detail/queue_impl.hpp

```diff
@@ -151,13 +151,16 @@ class queue_impl {
 
     MQueues.push_back(pi::cast<RT::PiQueue>(PiQueue));
 
-    RT::PiDevice Device{};
+    RT::PiDevice DevicePI{};
     const detail::plugin &Plugin = getPlugin();
     // TODO catch an exception and put it to list of asynchronous exceptions
-    Plugin.call<PiApiKind::piQueueGetInfo>(MQueues[0], PI_QUEUE_INFO_DEVICE,
-                                           sizeof(Device), &Device, nullptr);
-    MDevice =
-        DeviceImplPtr(new device_impl(Device, Context->getPlatformImpl()));
+    Plugin.call<PiApiKind::piQueueGetInfo>(
+        MQueues[0], PI_QUEUE_INFO_DEVICE, sizeof(DevicePI), &DevicePI, nullptr);
+    MDevice = MContext->findMatchingDeviceImpl(DevicePI);
+    if (MDevice == nullptr)
+      throw sycl::exception(
+          make_error_code(errc::invalid),
+          "internal error.  Device on PiQueue not found in Context.");
```

Comment:
If we can't find pi_device in a context then it means that user provided Device to the make_queue API which is not in the Context provided to the make_queue as an argument. And it seems like it is not an internal error. But it looks like its better to check this (that the device exist the provided context) and throw in make_queue_impl if check fail and put an assert here.
What I am trying to say is that we probably should not  throw a sycl exception for an internal error and maybe this error is not internal.

Reply:
@sergey-semenov  - do you have an opinion on this? Artur and I were discussing it and comparing the other queue_impl constructors and couldn't clearly decide if an assert was better than an exception and/or if merely updating the exception message was all that was necessary.

Reply:
I think just changing the exception message makes sense here since this constructor is only supposed to be called by `make_queue` anyway.

Reply:
done